File: ios/RCTNendAdViewManager.m
//
//  RCTNendAdViewManager.m
//  AwesomeProject
//

#import "RCTNendAdViewManager.h"

#import <React/RCTView.h>
#import <React/RCTUIManager.h>

@import NendAd;

@interface RCTNADView : RCTView <NADViewDelegate>

@property (nonatomic, strong) NADView *adView;
@property (nonatomic, copy) NSString *spotId;
@property (nonatomic, copy) NSString *apiKey;
@property (nonatomic) BOOL adjustSize;
@property (nonatomic, copy) RCTBubblingEventBlock onAdLoaded;
@property (nonatomic, copy) RCTBubblingEventBlock onAdFailedToLoad;
@property (nonatomic, copy) RCTBubblingEventBlock onAdClicked;
@property (nonatomic, copy) RCTBubblingEventBlock onInformationClicked;

@end

@implementation RCTNADView

- (void)loadAd
{
  if (self.spotId && self.apiKey) {
    if (!self.adView) {
      self.adView = [[NADView alloc] initWithIsAdjustAdSize:self.adjustSize];
      [self addSubview:self.adView];
      self.adView.delegate = self;
      [self.adView setNendID:self.apiKey spotID:self.spotId];
      [self.adView load];
    }
  }
}

#pragma mark - NADViewDelegate

- (void)nadViewDidReceiveAd:(NADView *)adView
{
  if (self.onAdLoaded) {
    NSDictionary<NSString *, NSNumber *> *body = @{@"width": @(CGRectGetWidth(self.adView.bounds)),
                                                   @"height": @(CGRectGetHeight(self.adView.bounds))};
    self.onAdLoaded(body);
  }
}

- (void)nadViewDidFailToReceiveAd:(NADView *)adView
{
  if (self.onAdFailedToLoad) {
    NSError *error = adView.error;
    NSDictionary<NSString *, id> *body = @{@"code": @(error.code), @"message": error.domain};
    self.onAdFailedToLoad(body);
  }
}

- (void)nadViewDidClickAd:(NADView *)adView
{
  if (self.onAdClicked) {
    self.onAdClicked(nil);
  }
}

- (void)nadViewDidClickInformation:(NADView *)adView
{
  if (self.onInformationClicked) {
    self.onInformationClicked(nil);
  }
}

@end

@implementation RCTNendAdViewManager

RCT_EXPORT_MODULE()

- (UIView *)view
{
  return [RCTNADView new];
}

RCT_EXPORT_METHOD(loadAd:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *,UIView *> *viewRegistry) {
    RCTNADView *view = (RCTNADView *)viewRegistry[reactTag];
    if (view) {
      [view loadAd];
    }
  }];
}

RCT_EXPORT_METHOD(resume:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *,UIView *> *viewRegistry) {
    RCTNADView *view = (RCTNADView *)viewRegistry[reactTag];
    if (view) {
      [view.adView resume];
    }
  }];
}

RCT_EXPORT_METHOD(pause:(nonnull NSNumber *)reactTag)
{
  [self.bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *,UIView *> *viewRegistry) {
    RCTNADView *view = (RCTNADView *)viewRegistry[reactTag];
    if (view) {
      [view.adView pause];
    }
  }];
}

RCT_EXPORT_VIEW_PROPERTY(spotId, NSString)
RCT_EXPORT_VIEW_PROPERTY(apiKey, NSString)
RCT_EXPORT_VIEW_PROPERTY(adjustSize, BOOL)
RCT_EXPORT_VIEW_PROPERTY(onAdLoaded, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdFailedToLoad, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdClicked, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onInformationClicked, RCTBubblingEventBlock)

@end

File: ios/RCTNendAdViewManager.swift
//
//  RCTNendAdViewManager.swift
//  AwesomeProject
//

import UIKit
import NendAd

@objc(RCTNendAdViewManager)
class RCTNendAdViewManager: RCTViewManager {
  static override func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  override var methodQueue: DispatchQueue! {
    return DispatchQueue.main
  }
  
  override func view() -> UIView! {
    return RCTNADView()
  }
  
  @objc(loadAd:)
  func loadAd(reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (manager, viewRegistry) in
      if let view = viewRegistry?[reactTag] as? RCTNADView {
        view.loadAd()
      }
    }
  }

  @objc(resume:)
  func resume(reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (manager, viewRegistry) in
      if let view = viewRegistry?[reactTag] as? RCTNADView, let adView = view.adView {
        adView.resume()
      }
    }
  }

  @objc(pause:)
  func pause(reactTag: NSNumber) {
    bridge.uiManager.addUIBlock { (manager, viewRegistry) in
      if let view = viewRegistry?[reactTag] as? RCTNADView, let adView = view.adView {
        adView.pause()
      }
    }
  }
}

class RCTNADView: RCTView {
  var spotId: String?
  var apiKey: String?
  var adjustSize = false
  var onAdLoaded: RCTBubblingEventBlock?
  var onAdFailedToLoad: RCTBubblingEventBlock?
  var onAdClicked: RCTBubblingEventBlock?
  var onInformationClicked: RCTBubblingEventBlock?
  
  fileprivate var adView: NADView?
  
  fileprivate func loadAd() {
    guard let spotId = self.spotId, let apiKey = self.apiKey, adView == nil else { return }
    adView = NADView(isAdjustAdSize: adjustSize)
    addSubview(adView!)
    adView!.delegate = self
    adView!.setNendID(apiKey, spotID: spotId)
    adView!.load()
  }
}

extension RCTNADView: NADViewDelegate {
  func nadViewDidReceiveAd(_ adView: NADView!) {
    if let callback = onAdLoaded {
      callback(["width": adView.bounds.size.width, "height": adView.bounds.size.height])
    }
  }
  
  func nadViewDidFail(toReceiveAd adView: NADView!) {
    if let callback = onAdFailedToLoad {
      let error = adView.error as NSError
      let body: [String: Any] = ["code": error.code, "message": error.domain]
      callback(body)
    }
  }
  
  func nadViewDidClickAd(_ adView: NADView!) {
    if let callback = onAdClicked {
      callback(nil)
    }
  }
  
  func nadViewDidClickInformation(_ adView: NADView!) {
    if let callback = onInformationClicked {
      callback(nil)
    }
  }
}
